packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -174,10 +251,122 @@ contract FederatedAttestations is
+   */
+  function lookupAllAttestations(bytes32 identifier, address[] calldata trustedIssuers)
+    external
+    view
+    returns (uint256[] memory, address[] memory, uint256[] memory, address[] memory)
+  {
+    uint256 totalAttestations;
+    uint256[] memory countsPerIssuer;
+
+    (totalAttestations, countsPerIssuer) = getTotalNumberOfAttestations(identifier, trustedIssuers);
+    address[] memory accounts = new address[](totalAttestations);
+    uint256[] memory issuedOns = new uint256[](totalAttestations);
+    address[] memory signers = new address[](totalAttestations);
+
+    OwnershipAttestation[] memory attestationsPerIssuer;
+    // Reset this and use as current index to get around stack-too-deep
+    totalAttestations = 0;
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      attestationsPerIssuer = identifierToAddresses[identifier][trustedIssuers[i]];
+      for (uint256 j = 0; j < attestationsPerIssuer.length; j = j.add(1)) {
+        accounts[totalAttestations] = attestationsPerIssuer[j].account;
+        issuedOns[totalAttestations] = attestationsPerIssuer[j].issuedOn;
+        signers[totalAttestations] = attestationsPerIssuer[j].signer;
+        totalAttestations = totalAttestations.add(1);
+      }
+    }
+    return (countsPerIssuer, accounts, issuedOns, signers);
+  }
+
+  /**
+   * @notice Calculates the total number of identifiers completed for an identifier
+             by each trusted issuer, from unrevoked signers only
+             This can be used to obtain an exact value for maxIdentifiers
+             prior to calling lookupIdentifiersByAddress
+   * @param account Address of the account
+   * @param trustedIssuers Array of n issuers whose identifiers will be included
+   * @return [0] Sum total of identifiers found
+   *         [1] Array of number of identifiers found per issuer
+   */
+  function getNumberOfUnrevokedIdentifiers(address account, address[] calldata trustedIssuers)
+    external
+    view
+    returns (uint256, uint256[] memory)
+  {
+    uint256 totalIdentifiers = 0;
+    uint256[] memory countsPerIssuer = new uint256[](trustedIssuers.length);
+
+    OwnershipAttestation[] memory attestationsPerIssuer;
+    bytes32[] memory identifiersPerIssuer;
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      identifiersPerIssuer = addressToIdentifiers[account][trustedIssuers[i]];
+      for (uint256 j = 0; j < identifiersPerIssuer.length; j = j.add(1)) {
+        bytes32 identifier = identifiersPerIssuer[j];
+        // Check if the mapping was produced by a revoked signer
+        attestationsPerIssuer = identifierToAddresses[identifier][trustedIssuers[i]];
+        for (uint256 k = 0; k < attestationsPerIssuer.length; k = k.add(1)) {
+          OwnershipAttestation memory attestation = attestationsPerIssuer[k];
+          // (identifier, account, issuer) tuples should be unique
```

Comment:
We should eventually add a comment here and in other places to point readers to where they can learn more about why this is the case.